File: nodejs/middleware/sqlInjection.js
'use strict';

const { scanSources } = require('../utils/patternMatcher');
const { logBlock } = require('../utils/logger');

const SQL_RULES = [
  // Critical — immediate block
  { name: 'sql-union-select',     severity: 'critical', pattern: /\bunion\s+(?:all\s+)?select\b/i },
  { name: 'sql-drop-table',       severity: 'critical', pattern: /;\s*drop\s+table\b/i },
  { name: 'sql-xp-cmdshell',      severity: 'critical', pattern: /\bxp_cmdshell\b/i },
  { name: 'sql-exec',             severity: 'critical', pattern: /\bexec(?:ute)?\s*\(/i },
  { name: 'sql-information-schema', severity: 'critical', pattern: /\binformation_schema\b/i },
  { name: 'sql-sleep',            severity: 'critical', pattern: /\bsleep\s*\(\s*\d/i },
  { name: 'sql-benchmark',        severity: 'critical', pattern: /\bbenchmark\s*\(/i },
  { name: 'sql-load-file',        severity: 'critical', pattern: /\bload_file\s*\(/i },
  { name: 'sql-into-outfile',     severity: 'critical', pattern: /\binto\s+(?:out|dump)file\b/i },
  { name: 'sql-sys-tables',       severity: 'critical', pattern: /\bsysobjects\b|\bsyscolumns\b/i },

  // High
  { name: 'sql-comment',          severity: 'high',     pattern: /(?:--|\/\*|\*\/|#\s*$)/m },
  { name: 'sql-stacked-query',    severity: 'high',     pattern: /;\s*(?:select|insert|update|delete|drop|alter|create|exec)\b/i },
  { name: 'sql-cast-convert',     severity: 'high',     pattern: /\b(?:cast|convert)\s*\(/i },
  { name: 'sql-char-concat',      severity: 'high',     pattern: /\bchar\s*\(\s*\d/i },

  // Medium
  { name: 'sql-boolean-true',     severity: 'medium',   pattern: /\bor\s+['"0-9]+\s*=\s*['"0-9]+/i },
  { name: 'sql-boolean-and',      severity: 'medium',   pattern: /\band\s+['"0-9]+\s*=\s*['"0-9]+/i },
  { name: 'sql-order-by-num',     severity: 'medium',   pattern: /\border\s+by\s+\d+\b/i },
  { name: 'sql-tautology',        severity: 'medium',   pattern: /'\s*or\s*'[^']*'\s*=\s*'/i },
  // Additional coverage
  { name: 'sql-waitfor-delay',    severity: 'critical', pattern: /\bwaitfor\s+delay\b/i },
  { name: 'sql-pg-sleep',         severity: 'critical', pattern: /\bpg_sleep\s*\(/i },
  { name: 'sql-hex-values',       severity: 'medium',   pattern: /0x[0-9a-f]{4,}/i },
  { name: 'sql-group-by-having',  severity: 'medium',   pattern: /\bhaving\s+\d+\s*=\s*\d+/i },
  { name: 'sql-dbms-fingerprint', severity: 'medium',   pattern: /\b(?:@@version|version\s*\(\s*\)|user\s*\(\s*\)|database\s*\(\s*\))\b/i },
  { name: 'sql-declare-set',      severity: 'high',     pattern: /\bdeclare\s+@\w+\b/i },
  { name: 'sql-bulk-insert',      severity: 'critical', pattern: /\bbulk\s+insert\b/i },
  { name: 'sql-openrowset',       severity: 'critical', pattern: /\bopenrowset\s*\(/i },
];

function createSqlInjectionMiddleware(config) {
  return function sqlInjectionMiddleware(req, res, next) {
    if (req.wafTrusted) return next();

    const sources = [
      { label: 'query', data: req.query },
      { label: 'body',  data: req.body },
      { label: 'path',  data: req.path },
      { label: 'cookies', data: req.cookies },
    ];

    const hit = scanSources(sources, SQL_RULES);

    if (hit) {
      const ip = req.wafIp || req.socket?.remoteAddress || 'unknown';
      const ruleDef = SQL_RULES.find((r) => r.name === hit.rule);

      logBlock({
        logPath: config.logPath,
        ip,
        method: req.method,
        path: req.path,
        rule: hit.rule,
        matched: hit.matched,
        source: hit.source,
        severity: ruleDef?.severity || 'medium',
        userAgent: req.headers['user-agent'] || '',
      });

      if (config.mode === 'log-only') return next();

      return res.status(403).json({
        blocked: true,
        rule: hit.rule,
        message: 'Request blocked by WAF',
      });
    }

    next();
  };
}

module.exports = createSqlInjectionMiddleware;

File: python/firewtwall/detectors/sql_injection.py
from __future__ import annotations
import re
from typing import Dict, Optional
from ._base import DetectorResult, _scan

_F = re.IGNORECASE

_RULES = [
    ("sqli-union-select",   "critical", re.compile(r"union\s+(?:all\s+)?select\b", _F)),
    ("sqli-or-tautology",   "critical", re.compile(r"\bor\s+\d+\s*=\s*\d+", _F)),
    ("sqli-and-tautology",  "critical", re.compile(r"\band\s+\d+\s*=\s*\d+", _F)),
    ("sqli-sleep",          "high",     re.compile(r"\b(?:sleep|benchmark|waitfor\s+delay)\s*\(", _F)),
    ("sqli-drop",           "critical", re.compile(r"\bdrop\s+(?:table|database|schema)\b", _F)),
    ("sqli-insert",         "high",     re.compile(r"\binsert\s+into\b", _F)),
    ("sqli-update-set",     "high",     re.compile(r"\bupdate\s+\w+\s+set\b", _F)),
    ("sqli-delete-from",    "high",     re.compile(r"\bdelete\s+from\b", _F)),
    ("sqli-exec",           "critical", re.compile(r"\bexec(?:ute)?\s*\(", _F)),
    ("sqli-xp-cmdshell",    "critical", re.compile(r"\bxp_cmdshell\b", _F)),
    ("sqli-information-schema","high",  re.compile(r"\binformation_schema\b", _F)),
    ("sqli-sys-tables",     "high",     re.compile(r"\bsys(?:objects|tables|columns)\b", _F)),
    ("sqli-cast",           "medium",   re.compile(r"\bcast\s*\(.*\bas\b", _F)),
    ("sqli-convert",        "medium",   re.compile(r"\bconvert\s*\(", _F)),
    ("sqli-char-concat",    "medium",   re.compile(r"\bchar\s*\(\s*\d+", _F)),
    ("sqli-comment-bypass", "high",     re.compile(r"(?:--|#|/\*)", _F)),
    ("sqli-stacked",        "high",     re.compile(r";\s*(?:select|insert|update|delete|drop)\b", _F)),
    ("sqli-hex-value",      "medium",   re.compile(r"\b0x[0-9a-f]{4,}\b", _F)),
    ("sqli-load-file",      "critical", re.compile(r"\bload_file\s*\(", _F)),
    ("sqli-into-outfile",   "critical", re.compile(r"\binto\s+(?:outfile|dumpfile)\b", _F)),
    ("sqli-schema-dump",    "high",     re.compile(r"\bschema\s*\(\)", _F)),
    ("sqli-group-by-having","medium",   re.compile(r"\bhaving\s+\d+\s*=\s*\d+", _F)),
    ("sqli-extractvalue",   "high",     re.compile(r"\bextractvalue\s*\(", _F)),
    ("sqli-updatexml",      "high",     re.compile(r"\bupdatexml\s*\(", _F)),
    ("sqli-floor-rand",     "high",     re.compile(r"\bfloor\s*\(\s*rand\s*\(", _F)),
    ("sqli-error-based",    "high",     re.compile(r"\b(?:exp|ln|log)\s*\(\s*~", _F)),
    ("sqli-substring",      "medium",   re.compile(r"\bsubstr(?:ing)?\s*\(", _F)),
    ("sqli-ascii",          "medium",   re.compile(r"\bascii\s*\(", _F)),
    ("sqli-if-blind",       "high",     re.compile(r"\bif\s*\(\s*\d+\s*=\s*\d+", _F)),
    ("sqli-case-blind",     "high",     re.compile(r"\bcase\s+when\s+\d+\s*=\s*\d+", _F)),
    ("sqli-concat",         "medium",   re.compile(r"\bconcat\s*\(", _F)),
    ("sqli-group-concat",   "medium",   re.compile(r"\bgroup_concat\s*\(", _F)),
    ("sqli-order-by-num",   "medium",   re.compile(r"\border\s+by\s+\d+", _F)),
]


def scan(sources: Dict[str, str]) -> Optional[DetectorResult]:
    return _scan(_RULES, sources)


class SqlInjectionDetector:
    @staticmethod
    def scan(sources: Dict[str, str]) -> Optional[DetectorResult]:
        return scan(sources)
